### PR TITLE
Add/appgw-listener-waf

### PR DESCRIPTION
**Pull request by djackson-ans**
# Description

Added support for per listener and per path rule based WAF policies

## Type of change

Please select the option that is relevant.

  - [ ] Bug fix (non-breaking change which fixes an issue)
  - [X] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] This change requires a documentation update

# Checklist:

  - [X] My code follows the style guidelines of this project
  - [X] I have performed a self-review of my code
  - [X] I have commented my code, particularly in hard-to-understand areas
  - [X] I have made corresponding changes to the documentation
  - [X] My changes generate no new warnings e.g. code analysis tooling or general use of the changed code

### CODE REVIEW

File: azure/application-gateway/advanced/variables.tf [hunk 1]
```diff
@@ -369,4 +369,119 @@ variable "waf_configuration" {
     })), {})
   })
   default = null
+}
+
+variable "listener_waf_configuration" {
+  description = "Defining the WAF policy per listener"
+  type = map(object({ // Key = policy name
+    associated_listeners = list(string)
+    firewall_mode        = optional(string, "Prevention")
```
Comment (jk-ans) [inline azure/application-gateway/advanced/variables.tf:378]: should this not be detection by default?

File: azure/application-gateway/advanced/variables.tf [hunk 2]
```diff
@@ -369,4 +369,119 @@ variable "waf_configuration" {
...
+    file_upload_limit_mb     = optional(number, 500)
+    max_request_body_size_kb = optional(number, 128)
+    managed_rule_exclusion = optional(list(object({
+      match_variable          = string
+      selector_match_operator = optional(string)
+      selector                = optional(string)
+    })), [])
+    custom_rules = optional(map(object({
+      priority = number
+      action   = optional(string, "Block")
```
Comment (jk-ans) [inline azure/application-gateway/advanced/variables.tf:403]: should this be Log by default?

File: azure/application-gateway/advanced/variables.tf [hunk 3]
```diff
@@ -369,4 +369,119 @@ variable "waf_configuration" {
...
+    )
+    error_message = "A listener is defined in multiple policies. This module only supports assigning a listener to a single policy."
+  }
+}
+
+variable "path_rule_waf_configuration" {
+  description = "Defining the WAF policy per path rule"
+  type = map(object({ // Key = policy name
+    associated_path_rules = list(string)
+    firewall_mode         = optional(string, "Prevention")
```
Comment (jk-ans) [inline azure/application-gateway/advanced/variables.tf:436]: should this not be detection by default?

File: azure/application-gateway/advanced/variables.tf [hunk 4]
```diff
@@ -369,4 +369,119 @@ variable "waf_configuration" {
...
+    file_upload_limit_mb     = optional(number, 500)
+    max_request_body_size_kb = optional(number, 128)
+    managed_rule_exclusion = optional(list(object({
+      match_variable          = string
+      selector_match_operator = optional(string)
+      selector                = optional(string)
+    })), [])
+    custom_rules = optional(map(object({
+      priority = number
+      action   = optional(string, "Block")
```
Comment (jk-ans) [inline azure/application-gateway/advanced/variables.tf:461]: should this not be Log by default?

File: azure/application-gateway/advanced/variables.tf [hunk 5]
```diff
@@ -369,4 +369,119 @@ variable "waf_configuration" {
...
+}
+
+variable "path_rule_waf_configuration" {
+  description = "Defining the WAF policy per path rule"
+  type = map(object({ // Key = policy name
+    associated_path_rules = list(string)
+    firewall_mode         = optional(string, "Prevention")
+
+    enable_OWASP           = optional(bool, true)
+    OWASP_rule_set_version = optional(string, "3.2")
```
Comment (jk-ans) [inline azure/application-gateway/advanced/variables.tf:439]: Latest version is 3.3.5

File: azure/application-gateway/advanced/variables.tf [hunk 6]
```diff
@@ -369,4 +369,119 @@ variable "waf_configuration" {
...
+}
+
+variable "listener_waf_configuration" {
+  description = "Defining the WAF policy per listener"
+  type = map(object({ // Key = policy name
+    associated_listeners = list(string)
+    firewall_mode        = optional(string, "Prevention")
+
+    enable_OWASP           = optional(bool, true)
+    OWASP_rule_set_version = optional(string, "3.2")
```
Comment (jk-ans) [inline azure/application-gateway/advanced/variables.tf:381]: latest version is 3.3.5

File: azure/application-gateway/advanced/variables.tf (hunk 1 above)
Comment (djackson-ans) [reply azure/application-gateway/advanced/variables.tf:378]: Most LLD's request prevention was the thinking here

File: azure/application-gateway/advanced/variables.tf (hunk 2 above)
Comment (djackson-ans) [reply azure/application-gateway/advanced/variables.tf:403]: Most LLD's that I've come across have (when a custom rule is specified) specified that Block is the action. 

If you believe this should be more explicit, then perhaps we remove the "optional" variable entirely and make it mandatory?

File: azure/application-gateway/advanced/variables.tf (hunk 4 above)
Comment (djackson-ans) [reply azure/application-gateway/advanced/variables.tf:461]: See above

File: azure/application-gateway/advanced/variables.tf (hunk 5 above)
Comment (djackson-ans) [reply azure/application-gateway/advanced/variables.tf:439]: TF Limitation:

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/web_application_firewall_policy#version

File: azure/application-gateway/advanced/variables.tf (hunk 6 above)
Comment (djackson-ans) [reply azure/application-gateway/advanced/variables.tf:381]: TF Limitation:

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/web_application_firewall_policy#version

**Review by jk-ans — APPROVED** [review]

**Pull request merged**